{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "602d2ed9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50656a34",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_card = \"../temp/data/raw/card.20250819.json\"\n",
    "\n",
    "with open(path_to_card, \"r\") as f:\n",
    "    card_data: dict[str, dict] = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfebbf0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# {\n",
    "#     \"model_id\": \"45\",\n",
    "#     \"model_name\": \"mdtP\",\n",
    "#     \"model_type\": \"protein homolog model\",\n",
    "#     \"model_type_id\": \"40292\",\n",
    "#     \"model_description\": \"Protein Homolog Models (PHM) detect protein sequences based on their similarity to a curated reference sequence, using curated BLASTP bitscore cut-offs. Protein Homolog Models apply to all genes that confer resistance through their presence in an organism, such as the presence of a beta-lactamase gene on a plasmid. PHMs include a reference sequence and a bitscore cut-off for detection using BLASTP. A Perfect RGI match is 100% identical to the reference protein sequence along its entire length, a Strict RGI match is not identical but the bit-score of the matched sequence is greater than the curated BLASTP bit-score cutoff, Loose RGI matches have a bit-score less than the curated BLASTP bit-score cut-off.\",\n",
    "#     \"model_param\": {\n",
    "#         \"blastp_bit_score\": {\n",
    "#             \"param_type\": \"BLASTP bit-score\",\n",
    "#             \"param_description\": \"The BLASTP bit-score is a numerical value that describes the overall quality of an alignment. Higher numbers correspond to higher similarity. The bit-score (S) is determined by the following formula: S = (λ × S − lnK)/ ln2 where λ is the Gumble distribution constant, S is the raw alignment score, and K is a constant associated with the scoring matrix. This parameter is used by AMR detection models that use a protein reference sequence, e.g. the protein homolog model. The BLASTP bit-score parameter is a curated value determined from BLASTP analysis of the canonical reference sequence of a specific AMR-associated protein against the database of CARD reference sequence. This value establishes a threshold for computational prediction of RGI Strict match (above bit-score cut-off) or Loose match (below bit-score cut-off).\",\n",
    "#             \"param_type_id\": \"40725\",\n",
    "#             \"param_value\": \"875\",\n",
    "#         }\n",
    "#     },\n",
    "#     \"model_sequences\": {\n",
    "#         \"sequence\": {\n",
    "#             \"4454\": {\n",
    "#                 \"protein_sequence\": {\n",
    "#                     \"accession\": \"BAE78082.1\",\n",
    "#                     \"sequence\": \"MINRQLSRLLLCSILGSTTLISGCALVRKDSAPHQQLKPEQIKLADDIHLASSGWPQAQWWKQLNDPQLDALIQRTLSGSHTLAEAKLREEKAQSQADLLDAGSQLQVAALGMLNRQRVSANGFLSPYSMDAPALGMDGPYYTEATVGLFAGLDLDLWGVHRSAVAAAIGAHNAALAETAAVELSLATGVAQLYYSMQASYQMLDLLEQTHDVIDYAVKAHQSKVAHGLEAQVPFHGARAQILAVDKQIVAVKGQITETRESLRALIGAGASDMPEIRPVALPQVQTGIPATLSYELLARRPDLQAMRWYVQASLDQVDSARALFYPSFDIKAFFGLDSIHLHTLFKKTSRQFNFIPGLKLPLFDGGRLNANLEGTRAASNMMIERYNQSVLNAVRDVAVNGTRLQTLNDEREMQAERVEATRFTQRAAEAAYQRGLTSRLQATEARLPVLAEEMSLLMLDSRRVIQSIQLMKSLGGGYQAGPVVEKK\",\n",
    "#                 },\n",
    "#                 \"dna_sequence\": {\n",
    "#                     \"accession\": \"AP009048.1\",\n",
    "#                     \"fmin\": \"4303042\",\n",
    "#                     \"fmax\": \"4304509\",\n",
    "#                     \"strand\": \"-\",\n",
    "#                     \"sequence\": \"ATGATCAATCGTCAACTTTCACGTCTGCTGTTGTGCAGCATTCTCGGCAGCACGACGCTGATTTCCGGCTGTGCCCTGGTACGTAAGGATTCTGCACCTCATCAACAGCTCAAACCGGAACAAATCAAACTGGCCGACGATATTCATCTTGCCAGCTCCGGCTGGCCGCAGGCGCAGTGGTGGAAACAACTCAATGACCCGCAGCTGGATGCGCTGATCCAACGGACGCTAAGTGGTTCACACACCCTCGCCGAAGCGAAACTGCGGGAAGAAAAAGCGCAGTCCCAGGCCGATTTGTTAGATGCCGGTTCACAATTACAGGTCGCAGCGTTAGGGATGCTCAACCGCCAACGTGTCTCGGCGAACGGCTTTTTAAGCCCTTATTCAATGGATGCGCCAGCACTGGGTATGGACGGGCCGTACTATACGGAAGCCACAGTAGGTTTGTTTGCCGGACTGGATCTTGATTTGTGGGGTGTGCATCGCTCAGCGGTTGCCGCCGCCATTGGCGCGCATAATGCCGCGCTGGCAGAAACCGCAGCAGTAGAGCTATCGCTGGCCACGGGCGTAGCGCAGCTTTATTACAGTATGCAGGCCAGCTATCAGATGCTCGATCTGTTAGAACAAACTCACGATGTGATTGATTACGCGGTGAAAGCGCACCAGAGTAAAGTGGCGCACGGTCTGGAAGCGCAAGTGCCTTTCCACGGCGCGCGGGCACAGATTCTGGCGGTCGATAAACAAATTGTTGCCGTCAAAGGGCAAATCACCGAAACGCGAGAATCTCTGCGTGCATTGATTGGCGCGGGAGCCAGCGATATGCCGGAGATCAGACCGGTGGCATTACCGCAAGTCCAGACCGGCATTCCGGCGACACTCTCTTATGAGTTGCTCGCCAGACGCCCGGATCTGCAAGCCATGCGCTGGTATGTTCAGGCGTCATTAGATCAGGTGGATTCCGCGCGGGCGTTGTTCTATCCGAGCTTTGATATCAAAGCGTTTTTCGGTCTGGACTCCATCCATCTGCATACCTTATTCAAAAAAACCAGTCGCCAGTTCAACTTCATCCCGGGTCTGAAATTGCCGCTGTTTGACGGTGGACGGTTGAATGCCAATCTCGAAGGCACGCGCGCCGCCAGCAACATGATGATTGAACGTTACAACCAGTCAGTACTGAACGCGGTGCGTGACGTTGCCGTCAACGGCACGCGTCTGCAAACGCTCAACGACGAGCGAGAAATGCAGGCTGAACGCGTGGAAGCCACGCGCTTTACCCAGCGCGCTGCCGAGGCCGCCTATCAGCGCGGCTTAACCAGCCGCTTACAGGCCACCGAAGCCCGGTTGCCAGTGCTTGCCGAAGAGATGTCATTACTGATGCTGGACAGCCGCCGGGTGATCCAAAGCATTCAGTTGATGAAATCGCTGGGCGGCGGGTATCAGGCAGGTCCCGTCGTCGAGAAAAAATAA\",\n",
    "#                     \"partial\": \"0\",\n",
    "#                 },\n",
    "#                 \"NCBI_taxonomy\": {\n",
    "#                     \"NCBI_taxonomy_cvterm_id\": \"36839\",\n",
    "#                     \"NCBI_taxonomy_name\": \"Escherichia coli str. K-12 substr. W3110\",\n",
    "#                     \"NCBI_taxonomy_id\": \"316407\",\n",
    "#                 },\n",
    "#             }\n",
    "#         }\n",
    "#     },\n",
    "#     \"ARO_accession\": \"3003550\",\n",
    "#     \"ARO_id\": \"40152\",\n",
    "#     \"ARO_name\": \"mdtP\",\n",
    "#     \"CARD_short_name\": \"mdtP\",\n",
    "#     \"ARO_description\": \"Multidrug resistance efflux pump. Could be involved in resistance to puromycin, acriflavine and tetraphenylarsonium chloride.\",\n",
    "#     \"ARO_category\": {\n",
    "#         \"36003\": {\n",
    "#             \"category_aro_accession\": \"0010002\",\n",
    "#             \"category_aro_cvterm_id\": \"36003\",\n",
    "#             \"category_aro_name\": \"major facilitator superfamily (MFS) antibiotic efflux pump\",\n",
    "#             \"category_aro_description\": \"Directed pumping of antibiotic out of a cell to confer resistance. Major facilitator superfamily (MFS) transporters and ABC transporters comprise the two largest and most functionally diverse of the transporter superfamilies. However, MFS transporters are distinct from ABC transporters in both their primary sequence and structure and in the mechanism of energy coupling. As secondary transporters they are, like RND and SMR transporters, energized by the electrochemical proton gradient.\",\n",
    "#             \"category_aro_class_name\": \"AMR Gene Family\",\n",
    "#         },\n",
    "#         \"35963\": {\n",
    "#             \"category_aro_accession\": \"0000045\",\n",
    "#             \"category_aro_cvterm_id\": \"35963\",\n",
    "#             \"category_aro_name\": \"acriflavine\",\n",
    "#             \"category_aro_description\": \"Acriflavine is a topical antiseptic. It has the form of an orange or brown powder. It may be harmful in the eyes or if inhaled. Acriflavine is also used as treatment for external fungal infections of aquarium fish.\",\n",
    "#             \"category_aro_class_name\": \"Antibiotic\",\n",
    "#         },\n",
    "#         \"35965\": {\n",
    "#             \"category_aro_accession\": \"0000047\",\n",
    "#             \"category_aro_cvterm_id\": \"35965\",\n",
    "#             \"category_aro_name\": \"puromycin\",\n",
    "#             \"category_aro_description\": \"Puromycin is an aminonucleoside antibiotic, derived from Streptomyces alboniger, that causes premature chain termination during ribosomal protein translation.\",\n",
    "#             \"category_aro_class_name\": \"Antibiotic\",\n",
    "#         },\n",
    "#         \"36174\": {\n",
    "#             \"category_aro_accession\": \"3000034\",\n",
    "#             \"category_aro_cvterm_id\": \"36174\",\n",
    "#             \"category_aro_name\": \"nucleoside antibiotic\",\n",
    "#             \"category_aro_description\": \"Nucleoside antibiotics are made of modified nucleosides and nucleotides with wide-ranging activities and means of antibacterial effects. This drug class includes aminonucleoside antibiotics, which contain an amino group.\",\n",
    "#             \"category_aro_class_name\": \"Drug Class\",\n",
    "#         },\n",
    "#         \"43746\": {\n",
    "#             \"category_aro_accession\": \"3005386\",\n",
    "#             \"category_aro_cvterm_id\": \"43746\",\n",
    "#             \"category_aro_name\": \"disinfecting agents and antiseptics\",\n",
    "#             \"category_aro_description\": \"Disinfectants that can also interact with antimicrobial resistance mechanisms, e.g. molecule efflux, and thus are the targets of disinfectant resistance.\",\n",
    "#             \"category_aro_class_name\": \"Drug Class\",\n",
    "#         },\n",
    "#         \"36298\": {\n",
    "#             \"category_aro_accession\": \"3000159\",\n",
    "#             \"category_aro_cvterm_id\": \"36298\",\n",
    "#             \"category_aro_name\": \"efflux pump complex or subunit conferring antibiotic resistance\",\n",
    "#             \"category_aro_description\": \"Efflux proteins that pump antibiotic out of a cell to confer resistance.\",\n",
    "#             \"category_aro_class_name\": \"Efflux Component\",\n",
    "#         },\n",
    "#         \"36001\": {\n",
    "#             \"category_aro_accession\": \"0010000\",\n",
    "#             \"category_aro_cvterm_id\": \"36001\",\n",
    "#             \"category_aro_name\": \"antibiotic efflux\",\n",
    "#             \"category_aro_description\": \"Antibiotic resistance via the transport of antibiotics out of the cell.\",\n",
    "#             \"category_aro_class_name\": \"Resistance Mechanism\",\n",
    "#         },\n",
    "#     },\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8189120d",
   "metadata": {},
   "outputs": [],
   "source": [
    "keys_to_pop = ['_version', '_comment', '_timestamp']\n",
    "for key in keys_to_pop:\n",
    "    card_data.pop(key, None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b778678",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get values for specific keys\n",
    "data_rows = []\n",
    "for record in card_data.values():\n",
    "    protein_sequence = \"\"\n",
    "    dna_sequence = \"\"\n",
    "    antibiotics = []\n",
    "    drug_classes = []\n",
    "    resistance_mechanisms = []\n",
    "    amr_gene_families = []\n",
    "\n",
    "    try:\n",
    "        for value in record[\"model_sequences\"][\"sequence\"].values():\n",
    "            protein_sequence = value[\"protein_sequence\"][\"sequence\"]\n",
    "            dna_sequence = value[\"dna_sequence\"][\"sequence\"]\n",
    "\n",
    "        for value in record[\"ARO_category\"].values():\n",
    "            category_name = value[\"category_aro_name\"]\n",
    "            category_class_name = value[\"category_aro_class_name\"]\n",
    "            if category_class_name == \"Antibiotic\":\n",
    "                antibiotics.append(category_name)\n",
    "            elif category_class_name == \"Drug Class\":\n",
    "                drug_classes.append(category_name)\n",
    "            elif category_class_name == \"Resistance Mechanism\":\n",
    "                resistance_mechanisms.append(category_name)\n",
    "            elif category_class_name == \"AMR Gene Family\":\n",
    "                amr_gene_families.append(category_name)\n",
    "    except KeyError:\n",
    "        print(f\"Key not found in record {record[\"model_id\"]}, skipping...\")\n",
    "        continue\n",
    "\n",
    "    data_rows.append(\n",
    "        {\n",
    "            \"protein_sequence\": protein_sequence,\n",
    "            \"dna_sequence\": dna_sequence,\n",
    "            \"antibiotics\": antibiotics,\n",
    "            \"drug_classes\": drug_classes,\n",
    "            \"resistance_mechanisms\": resistance_mechanisms,\n",
    "            \"amr_gene_families\": amr_gene_families,\n",
    "        }\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b870695",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Extracted {len(data_rows)} records from CARD data.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b720506e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pl.DataFrame(data_rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "687916dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f301aa35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Put it in ../temp/data/interim/card_extracted.parquet\n",
    "df.write_parquet(\"../temp/data/interim/card_extracted.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee8ad66c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract drug classes, and one hot encode them for future analysis\n",
    "\n",
    "# explode and pivot\n",
    "\n",
    "\n",
    "df_exploded = df.drop([\"antibiotics\", \"resistance_mechanisms\", \"amr_gene_families\"]).explode(\"drug_classes\")\n",
    "df_one_hot = df_exploded.with_columns(\n",
    "    pl.col(\"drug_classes\").cast(pl.Categorical)\n",
    ").to_dummies(columns=[\"drug_classes\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0371c35b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_one_hot = df_one_hot.rename(\n",
    "    {col: col.removeprefix(\"drug_classes_\") for col in df_one_hot.columns}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c1d2f9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ../temp/data/interim/card_amr.parquet\n",
    "df_one_hot.write_parquet(\"../temp/data/interim/card_amr.parquet\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
