# Idea 1
Processing a genome, we can extract ORFs (Open Reading Frames) which are sequences that could potentially code for proteins. We can then use TARP to predict the function (whether it's an AMR gene or not) of these ORFs based on raw nucleotide sequences.

# Challenge 1
How the model perform on compared to existing tools for AMR gene prediction?
## PLM-ARG:
PLM-ARG is a protein language model specifically designed for AMR gene prediction. It uses protein sequences as input and has shown high accuracy in identifying AMR genes. However, it requires the translation of nucleotide sequences to protein sequences, which can introduce errors if the ORF prediction is not accurate. Direct comparison of TARP's performance on nucleotide sequences with PLM-ARG's performance on protein sequences is hard due to the different input types.

### How can we compare PLM-ARG and TARP?
Since Nucleotide to Protein translation is not always straightforward. We can take a dataset of known AMR genes with both nucleotide and protein sequences available. We can then run TARP on the nucleotide sequences and PLM-ARG on the corresponding protein sequences. By comparing the predictions from both models against the known labels, we can evaluate their performance using metrics such as accuracy, precision, recall, and F1-score.

# Challenge 2
However, most models for AMR gene prediction are based on protein sequences rather than nucleotide sequences. This is because protein sequences provide more functional information, making it easier to identify AMR genes. Therefore, TARP's performance on nucleotide sequences may not be directly comparable to models that use protein sequences.

### Possible Solution
Tarp can be adopted for a translation step, where it first predicts the protein sequence from the nucleotide sequence and then uses a protein-based AMR prediction model. This two-step approach could leverage the strengths of both nucleotide and protein-based models.